File: solutions/Lowest Common Ancestor of a Binary Search Tree/Lowest Common Ancestor of a Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        
        # edge cases
            # empty BST ?

        # strategy
        # get ancestry paths for each node
        # find first LCA in each ancestry paths


        def get_path(node, target):
            arr = []

            while(node.val != target):
                arr.append(node.val)
                if node.val < target:
                    node = node.right
                elif node.val > target:
                    node = node.left
            
            arr.append(node.val)
            return arr

        p_path = get_path(root, p.val)
        q_path = get_path(root, q.val)

        # print(p_path)
        # print(q_path)

        LCA = root.val
        for i in range( min(len(p_path), len(q_path)) ):
            if( p_path[i] == q_path[i] ):
                LCA = (q_path[i])
            else:
                break
        
        return TreeNode(LCA)

File: solutions/Lowest Common Ancestor of a Binary Search Tree/Lowest Common Ancestor of a Binary Search Tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

 /**
GOAL: in a BST, find the closest parent/ancestor of 2 nodes

STRATEGY:
    for each target node,
        - create an ancestry list, from root to parent
    
    simultaneously iterate through both ancestry lists,
    until ancestors diverge

NOTES:
    - nodes can be their own parents
    
  */

function lowestCommonAncestor(root: TreeNode | null, p: TreeNode | null, q: TreeNode | null): TreeNode | null {
    const path1: TreeNode[] = [];
    const path2: TreeNode[] = [];

    createPath(root, path1, p);
    createPath(root, path2, q);

    let lca: TreeNode = root;
    let i: number = 0;
    while(i < Math.min(path1.length, path2.length) && path1[i].val === path2[i].val) {
        lca = path1[i];
        i++;
    }
    return lca;

    // func def
    function createPath(curNode: TreeNode | null, path: TreeNode[], targetNode: TreeNode): void {
        // base case
        if (curNode === null) return;
        else if (curNode.val === targetNode.val) {
            path.push(curNode);
            return;
        }

        // ops
        path.push(curNode);

        // recursion
        if (curNode.val < targetNode.val) createPath(curNode.right, path, targetNode);
        else if (curNode.val > targetNode.val) createPath(curNode.left, path, targetNode);
    }
};